pkg/analysis/passes/coderules/coderules.go

```diff
@@ -0,0 +1,153 @@
+package coderules
+
+import (
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/grafana/plugin-validator/pkg/analysis"
+	"github.com/grafana/plugin-validator/pkg/analysis/passes/sourcecode"
+	"github.com/grafana/plugin-validator/pkg/logme"
+)
+
+//go:embed semgrep-rules.yaml
+var semgrepRules string
+
+var (
+	codeRulesViolationError   = &analysis.Rule{Name: "code-rules-violation-error", Severity: analysis.Error}
+	codeRulesViolationWarning = &analysis.Rule{Name: "code-rules-violation-warning", Severity: analysis.Warning}
+	semgrepNotFound           = &analysis.Rule{Name: "semgrep-not-found", Severity: analysis.Warning}
+	semgrepRunningErr         = &analysis.Rule{Name: "semgrep-running-err", Severity: analysis.Warning}
+)
+
+var Analyzer = &analysis.Analyzer{
+	Name:     "code-rules",
+	Requires: []*analysis.Analyzer{sourcecode.Analyzer},
+	Run:      run,
+	Rules:    []*analysis.Rule{codeRulesViolationError, codeRulesViolationWarning, semgrepNotFound, semgrepRunningErr},
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	sourceCodeDir, ok := pass.ResultOf[sourcecode.Analyzer].(string)
+	if !ok {
+		// no source code for the validator
+		return nil, nil
+	}
+
+	// check if there's a pkg folder inside the source code
+	sourceCodePkgDir := filepath.Clean(fmt.Sprintf("%s/pkg/", sourceCodeDir))
+	info, err := os.Stat(sourceCodePkgDir)
+	// if there's a pkg folder, use it as the source code dir
+	if err == nil && info.IsDir() {
+		sourceCodeDir = sourceCodePkgDir
+	}
+
+	logme.DebugFln("source code dir: %s", sourceCodeDir)
+
+	semgrepPath, err := getSemgrepPath()
+	if err != nil {
+		// can't run semgrep
+		if semgrepNotFound.ReportAll {
+			pass.ReportResult(pass.AnalyzerName, semgrepNotFound, "semgrep not found in PATH", "")
+		}
+		return nil, nil
+	}
+
+	logme.DebugFln("semgrep path: %s", semgrepPath)
+
+	semgrepRulesPath, cleanup, err := getSemgrepRulesPath()
+	if err != nil {
+		// can't run semgrep
+		if semgrepNotFound.ReportAll {
```

Comment:
maybe this should be another rule called semgrep-rules-not-found instead of reusing the binary check?